tests/php/Domain/Services/CreateAccount.php

```diff
@@ -0,0 +1,236 @@
+
+		return [
+			'user_id' => $user_id,
+			'order' => $test_order,
+		];
+	}
+
+    /**
+     * Test successful user signup cases.
+     *
+     * @dataProvider create_customer_data
+     */
+	public function test_create_customer_from_order( $email, $first_name, $last_name, $options ) {
+		if ( ! TestedCreateAccount::is_feature_enabled() ) {
+            $this->markTestSkipped(
+              'Skipping CreateAccount test - experimental feature flag is disabled.'
+            );
+		}
+
+		$result = $this->execute_create_customer_from_order(
+			$email,
+			$first_name,
+			$last_name,
+			$options
+		);
+
+		$test_user = $this->factory()->user->get_object_by_id( $result['user_id'] );
+		$test_order = $result['order'];
+
+		$this->assertEquals( get_current_user_id(), $result['user_id'] );
+
+		$this->assertEquals( $test_user->first_name, $first_name );
+		$this->assertEquals( $test_user->last_name, $last_name );
+		$this->assertEquals( $test_user->user_email, $email );
+		$this->assertArraySubset( $test_user->roles, [ 'customer' ] );
+
+		$this->assertEquals( $test_order->get_customer_id(), $result['user_id'] );
+	}
+
+    public function create_customer_data()
+    {
+        return [
+        	// User requested an account.
+        	[
+        		'maryjones@testperson.net',
+        		'Mary',
+        		'Jones',
+				[
+					'should_create_account' => true,
+					'enable_guest_checkout' => true,
+				],
+        	],
+       	    // User requested an account + site doesn't allow guest.
+        	[
+        		'maryjones@testperson.net',
+        		'Mary',
+        		'Jones',
+				[
+					'should_create_account' => true,
+					'enable_guest_checkout' => false,
+				],
+        	],
+        	// User requested an account; name fields are not required.
+        	[
+        		'private_person@hotmail.com',
+        		'',
+        		'',
+				[
+					'should_create_account' => true,
+					'enable_guest_checkout' => true,
+				],
+        	],
+        	// Store does not allow guest - signup is required (automatic).
+        	[
+        		'henrykissinger@fbi.gov',
+        		'Henry',
+        		'Kissinger',
+				[
+					'should_create_account' => false,
+					'enable_guest_checkout' => false,
+				],
+        	],
+        ];
+    }
+
+    /**
+     * Test exception is thrown if user already signed up.
+     */
+	public function test_customer_already_exists() {
+		if ( ! TestedCreateAccount::is_feature_enabled() ) {
+            $this->markTestSkipped(
+              'Skipping CreateAccount test - experimental feature flag is disabled.'
+            );
+		}
+
+		$user_id = $this->factory()->user->create( [
+			'user_email' => 'maryjones@testperson.net',
+		] );
+
+		$this->expectException( \Exception::class );
+
+		$result = $this->execute_create_customer_from_order(
+    		'maryjones@testperson.net',
+    		'Mary',
+    		'Jones',
+			[
+				'should_create_account' => true,
+				'enable_guest_checkout' => true,
+			],
+		);
```

Comment:
Zoinks! Fixed here: 7a12383d25e6d207ba7aa8b1bb2cfe500045fa50

I see we exclude tests folder from phpcs linter, `<exclude-pattern>*/tests/*</exclude-pattern>`. I wonder if we can re-enable this to catch these automatically? I'm guessing there's a good reason. (I had a quick play and couldn't get the linter to find whitespace issues in tests files, maybe excluded in Woo or WP core rules?)

Reply:
🤔 I'm guessing it's to avoid some of the extra documentation around parameter definition, function comment blocks etc that is typically not needed for test files because the functions are typically more descriptive. We can look at expanding the lint coverage on test files for php too but that will surface a lot more linting fails needing handled, so probably good doing in separate issue.